src/Std/Data/DTreeMap/Internal/Impl/Query.lean

```diff
@@ -0,0 +1,319 @@
+  match t with
+  | .inner _ k' v' l r =>
+    match h : compare k k' with
+    | .lt => get k l (by simpa [contains, h] using hlk)
+    | .gt => get k r (by simpa [contains, h] using hlk)
+    | .eq => v'
+
+/-- Returns the value for the key `k`, or panics if such a key does not exist. -/
+def get! [Ord α] (k : α) (t : Impl α (fun _ => δ)) [Inhabited δ] : δ :=
+  match t with
+  | .leaf => panic! "Key is not present in map"
+  | .inner _ k' v' l r =>
+    match compare k k' with
+    | .lt => get! k l
+    | .gt => get! k r
+    | .eq => v'
+
+/-- Returns the value for the key `k`, or `fallback` if such a key does not exist. -/
+def getD [Ord α] (k : α) (t : Impl α (fun _ => δ)) (fallback : δ) : δ :=
+  match t with
+  | .leaf => fallback
+  | .inner _ k' v' l r =>
+    match compare k k' with
+    | .lt => getD k l fallback
+    | .gt => getD k r fallback
+    | .eq => v'
+
+end Const
+
+/-- The smallest element of `t` that is not less than `k`. Also known as `lookupGE` or `ceil`. -/
+@[inline]
+def lookupGE [Ord α] (k : α) : Impl α β → Option ((a : α) × β a) :=
+  go none
+where
+  go (best : Option ((a : α) × β a)) : Impl α β → Option ((a : α) × β a)
+  | .leaf => best
+  | .inner _ ky y l r => match compare k ky with
+    | .lt => go (some ⟨ky, y⟩) l
+    | .eq => some ⟨ky, y⟩
+    | .gt => go best r
+
+/-- The smallest element of `t` that is greater than `k`. Also known as `lookupGT` or `higher`. -/
+@[inline]
+def lookupGT [Ord α] (k : α) : Impl α β → Option ((a : α) × β a) :=
+  go none
+where
+  go (best : Option ((a : α) × β a)) : Impl α β → Option ((a : α) × β a)
+  | .leaf => best
+  | .inner _ ky y l r => match compare k ky with
+    | .lt => go (some ⟨ky, y⟩) l
+    | _ => go best r
+
+/-- The largest element of `t` that is not greater than `k`. Also known as `floor`. -/
+@[inline]
+def lookupLE [Ord α] (k : α) : Impl α β → Option ((a : α) × β a) :=
+  go none
+where
+  go (best : Option ((a : α) × β a)) : Impl α β → Option ((a : α) × β a)
+  | .leaf => best
+  | .inner _ ky y l r => match compare k ky with
+    | .lt => go best l
+    | .eq => some ⟨ky, y⟩
+    | .gt => go (some ⟨ky, y⟩) r
+
+/-- The largest element of `t` that is less than `k`. Also known as `lower`. -/
+@[inline]
+def lookupLT [Ord α] (k : α) : Impl α β → Option ((a : α) × β a) :=
+  go none
+where
+  go (best : Option ((a : α) × β a)) : Impl α β → Option ((a : α) × β a)
+  | .leaf => best
+  | .inner _ ky y l r => match compare k ky with
+    | .gt => go (some ⟨ky, y⟩) r
+    | _ => go best l
+
+/-- The smallest element of `t`. -/
+def min? [Ord α] : Impl α β → Option ((a : α) × β a)
+  | .leaf => none
+  | .inner _ k v .leaf _ => some ⟨k, v⟩
+  | .inner _ _ _ l@(.inner _ _ _ _ _) _ => l.min?
+
+/-- The largest element of `t`. -/
+def max? [Ord α] : Impl α β → Option ((a : α) × β a)
+  | .leaf => none
+  | .inner _ k v _ .leaf => some ⟨k, v⟩
+  | .inner _ _ _ _ r@(.inner _ _ _ _ _) => r.max?
+
```

Comment:
Is it not possible to write `atIndex` which takes a proof?

Reply:
I haven't touched these operations yet and they don't have a user-facing pendant yet. Will delete them for now.